extensions/EXT/EGL_EXT_platform_xcb.txt

```diff
@@ -0,0 +1,357 @@
+Name
+
+    EXT_platform_xcb
+
+Name Strings
+
+    EGL_EXT_platform_xcb
+
+Contributors
+
+    Yuxuan Shui <yshuiv7@gmail.com>
+
+Contacts
+
+    Yuxuan Shui <yshuiv7@gmail.com>
+
+Status
+
+    Complete
+
+Version
+
+    Version 1, 2020-08-28
+
+Number
+
+    EGL Extension #141
+
+Extension Type
+
+    EGL client extension
+
+Dependencies
+
+    Requires EGL_EXT_client_extensions to query its existence without
+    a display.
+
+    Requires EGL_EXT_platform_base.
+
+    This extension is written against the wording of version 9 of the
+    EGL_EXT_platform_base specification.
+
+Overview
+
+    This extension defines how to create EGL resources from native X11
+    resources using the functions defined by EGL_EXT_platform_base.
+
+    The native X11 resources required by this extension are xcb resources.
+    All X11 types discussed here are defined by the header `xcb.h`.
+
+New Types
+
+    None
+
+New Procedures and Functions
+
+    None
+
+New Tokens
+
+    Accepted as the <platform> argument of eglGetPlatformDisplayEXT:
+
+        EGL_PLATFORM_XCB_EXT                    0x31DC
+
+    Accepted as an attribute name in the <attrib_list> argument of
+    eglGetPlatformDisplayEXT:
+
+        EGL_PLATFORM_XCB_SCREEN_EXT             0x31DE
+
+Additions to the EGL Specification
+
+    None.
+
+New Behavior
+
+    To determine if the EGL implementation supports this extension, clients
+    should query the EGL_EXTENSIONS string of EGL_NO_DISPLAY.
+
+    This extension defines the same set of behaviors as the X11 platform
+    extension, except Xlib types are replaced with xcb types.
+
+    To obtain an EGLDisplay backed by an X11 screen, call
+    eglGetPlatformDisplayEXT with <platform> set to EGL_PLATFORM_XCB_EXT. The
+    <native_display> parameter specifies the X11 display connection to use, and
+    must point to a valid xcb `xcb_connection_t` or be NULL.  If <native_display>
+    is EGL_DEFAULT_DISPLAY, then EGL will create [1] a connection to the default
+    X11 display. The environment variable DISPLAY determines the default X11
+    display as described in the documentation of `xcb_connect`.  The value of
+    attribute EGL_PLATFORM_XCB_SCREEN_EXT specifies the X11 screen to use. If
+    the attribute is omitted from <attrib_list>, then the environment varialbe
+    DISPLAY will be used to select a screen.  Otherwise, the attribute's value
+    must be a valid screen on the display connection. If the attribute's value
+    is not a valid screen, then an EGL_BAD_ATTRIBUTE error is generated.
+
+    [fn1] The method by which EGL creates a connection to the default X11
```

Comment:
There may be a subtle interaction with EGL_NV_native_query here: Whatever method EGL uses, it must be capable of generating an xcb_connection_t to return to the app form it if EGL_NV_native_query is supported.

Reply:
This is the same language used in `EXT_platform_x11` too.

~~Mesa stores a `xcb_connection_t *` so this won't be a problem for them. If other drivers don't do this, they probably have to choose between not supporting `EXT_platform_xcb` or not supporting `NV_native_query`.~~

Oh wait, this extension doesn't define a `EGLNativeDisplayType`, which means `NV_native_query` have to create a Xlib connection from a `xcb_connection_t`.

I think it's probably best to update `NV_native_query` to say, the native display can only be retrieved if a `EGLNativeDisplayType` is used to create the `EGLDisplay`.

Reply:
I'm not clear why that is.  The native display type is a property of the EGLDisplay.  NV_native_query implicitly returns an object of the type corresponding to the display's native display type, right?

I agree this is likely an issue in EXT_platform_x11 as well, but I think simply noting an interaction as you originally defined above (If you do something crazy, you can't support both extensions) in the "dependencies" section would be fine to resolve this, and then we  could backport that to platform_x11.

Reply:
NV_native_query explicitly specifies that the function would return a `EGLNativeDisplayType`, which is the type passed to `eglGetDisplay`. This extension doesn't define a `EGLNativeDisplayType`.

Reply:
Yes, this is a rough edge, but I think in practice EGLNativeDisplayType will always be a pointer, and is effectively void* in any implementation that supports multiple platforms, so very pedantic C implementations (are there any C implementations where pointer types actually differ besides between function pointers and data pointers?) aside, I would assume it's just a matter of casting EGLNativeDisplayType back to the appropriate type for your platform.

Reply:
I can take this interpretation. So I guess this is a non-issue then?

Reply:
Pretty much. I think it's worth mentioning as an interaction issue (using text similar to suggested above), but if you disagree, not a big deal. It's a pretty dark corner case.